db/filters/base.py

```diff
@@ -0,0 +1,381 @@
+    type: LeafPredicateType = static(LeafPredicateType.LESSER_OR_EQUAL)
+    name: str = static("Lesser or equal")
+
+    def to_sa_filter(self):
+        return column(self.column) <= self.parameter
+
+
+@frozen_dataclass
+class Empty(NoParameter, Leaf):
+    type: LeafPredicateType = static(LeafPredicateType.EMPTY)
+    name: str = static("Empty")
+
+    def to_sa_filter(self):
+        return column(self.column) == None
+
+
+@frozen_dataclass
+class NotEmpty(NoParameter, Leaf):
+    type: LeafPredicateType = static(LeafPredicateType.NOT_EMPTY)
+    name: str = static("Not empty")
+
+    def to_sa_filter(self):
+        return column(self.column) != None
+
+
+@frozen_dataclass
+class In(MultiParameter, Leaf):
+    type: LeafPredicateType = static(LeafPredicateType.IN)
+    name: str = static("In")
+
+    def to_sa_filter(self):
+        return column(self.column).in_(self.parameters)
+
+
+@frozen_dataclass
+class NotIn(MultiParameter, Leaf):
+    type: LeafPredicateType = static(LeafPredicateType.NOT_IN)
+    name: str = static("Not in")
+
+    def to_sa_filter(self):
+        return column(self.column).not_in(self.parameters)
+
+
+@frozen_dataclass
+class Not(SingleParameter, Branch):
+    type: BranchPredicateType = static(BranchPredicateType.NOT)
+    name: str = static("Not")
+
+    def to_sa_filter(self):
+        return not_(*self.get_child_sa_filters())
+
+
+@frozen_dataclass
+class And(MultiParameter, Branch):
+    type: BranchPredicateType = static(BranchPredicateType.AND)
+    name: str = static("And")
+
+    def to_sa_filter(self):
+        return and_(*self.get_child_sa_filters())
+
+
+@frozen_dataclass
+class Or(MultiParameter, Branch):
+    type: BranchPredicateType = static(BranchPredicateType.OR)
+    name: str = static("Or")
+
+    def to_sa_filter(self):
+        return or_(*self.get_child_sa_filters())
+
+
+def get_predicate_subclass_by_type_str(predicate_type_str: str) -> Union[Type[LeafPredicateType], Type[BranchPredicateType]]:
+    for subclass in all_predicates:
+        if subclass.type.value == predicate_type_str:
+            return subclass
+    raise UnknownPredicateType(predicate_type_str)
+
+
+class BadFilterFormat(SABadFilterFormat):
+    pass
+
+
+class UnknownPredicateType(BadFilterFormat):
+    pass
+
+class ReferencedColumnsDontExist(BadFilterFormat):
+    pass
+
+
+all_predicates = [
+    Equal,
+    NotEqual,
+    Greater,
+    GreaterOrEqual,
+    Lesser,
+    LesserOrEqual,
+    Empty,
+    NotEmpty,
+    In,
+    NotIn,
+    Not,
+    And,
+    Or,
+]
+
+
+def _not_empty(xs):
+    return len(xs) > 0
+
+
+def _all_items_unique(xs):
+    for item1 in xs:
+        times_seen = 0
+        for item2 in xs:
+            if item1 == item2:
+                times_seen += 1
+            # A non-duplicate will be seen only once.
+            if times_seen == 2:
+                return False
+    return True
+
+
+def assert_predicate_correct(predicate):
```

Comment:
These can be broken up and put on the class definitions. That would however distribute the correctness definition across the file. The reason methods like this on or the deserialization method are large and nested is because they use control flow to traverse the predicate class hierarchy. I'm realising that that's a put off. As I said, the alternative is to distribute the correctness/deserialization definition (which I'm seriously considering).